src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java

```diff
@@ -288,6 +304,62 @@ public void channelCreated(Channel ch) {
     this.extraHttpHeaders = extraHttpHeaders;
     this.verifyDownloads = verifyDownloads;
     this.digestUtil = digestUtil;
+    this.retrier = retrier != null ? retrier : newRetrier(null);
+  }
+
+  public static RemoteRetrier newRetrier(RemoteOptions options) {
+    CircuitBreaker circuitBreaker = Retrier.ALLOW_ALL_CALLS;
+    if (options != null && options.remoteMaxFailures > 0) {
+      circuitBreaker = new CircuitBreaker() {
+        State state = State.ACCEPT_CALLS;
+        AtomicInteger failures = new AtomicInteger(0);
+        final int maxFailures = options.remoteMaxFailures;
+        @Override
+        public State state() {
+          return state;
+        }
+
+        @Override
+        public void recordFailure() {
+          if (failures.incrementAndGet() > maxFailures) {
+            state = State.REJECT_CALLS;
+          }
+        }
+
+        @Override
+        public void recordSuccess() { }
+      };
+    }
+
+    if (options == null || options.remoteMaxRetryAttempts <= 0) {
+      return new RemoteRetrier(
+          () -> Retrier.RETRIES_DISABLED,
+          (Exception e) -> false,
+          MoreExecutors.listeningDecorator(Executors.newScheduledThreadPool(1)),
+          circuitBreaker);
+    }
+    return new RemoteRetrier(
+        () -> new RemoteRetrier.ExponentialBackoff(options),
+        (Exception e) -> shouldRetry(e),
+        MoreExecutors.listeningDecorator(Executors.newScheduledThreadPool(1)),
+        circuitBreaker);
+  }
+
+  private static boolean shouldRetry(Exception e) {
```

Comment:
Which case is going to catch the `Connection reset by peer` errors?

Reply:
The `RETRYABLE_ERROR_MESSAGE` should catch it